### ISSUE TITLE
Using references

### ISSUE THREAD

**Issue by plusangel** [BlueBrain/HighFive#396]
Dear all,

first of all, I would like to thank the author and the contributors for this amazing work.
C++ API for HDF5 is not that straightforward and your work is absolutely amazing.
I had tried other libraries too...

I work in a library for NDT (Non-Destructive Testing) testing that uses HDF5 under the hood.
I have to use extensive references and I had a hard time.

My suggestion is to include the following example in the relevant directory, to help future users on that.
I took one of your example scripts and I turned it to use references, like that:

```cpp
#include <iostream>
#include <string>
#include <vector>

#include <highfive/H5DataSet.hpp>
#include <highfive/H5DataSpace.hpp>
#include <highfive/H5File.hpp>
#include <highfive/H5Reference.hpp>

using namespace HighFive;

const std::string FILE_NAME("dataset_integer.h5");
const std::string SOURCE_INT_DATASET_NAME("source_dateset");
const std::string REFERENCE_DATASET_NAME("reference_dateset");
const size_t size_dataset = 20;

// create a dataset 1D from a vector of string
void write_dataset() {

  // we create a new hdf5 file
  File file(FILE_NAME, File::ReadWrite | File::Create | File::Truncate);

  // we create a new group
  Group group = file.createGroup("a_group");

  std::vector<int> data(size_dataset);
  for (size_t i = 0; i < data.size(); ++i) {
    data[i] = int(i);
  }

  // let's create a dataset of native integer with the size of the vector
  // 'data' inside the group
  DataSet dataset = group.createDataSet<int>(SOURCE_INT_DATASET_NAME, DataSpace::From(data));
  dataset.write(data);

  // create a reference to the dateset containing the integers
  Reference ref = Reference(group, dataset);
  std::vector<Reference> ref_container{ref};

  // in similar fashion, we store as dataset the reference that we want
  DataSet ref_set = group.createDataSet<Reference>(REFERENCE_DATASET_NAME, DataSpace::From(ref_container));
  ref_set.write(ref_container);
}

// read our data back
void read_dataset() {
  // we open the existing hdf5 file we created before
  File file(FILE_NAME, File::ReadOnly);

  // we load the group
  Group my_group = file.getGroup("a_group");

  // we load the dataset that contains the reference
  DataSet ref_dataset = my_group.getDataSet(REFERENCE_DATASET_NAME);

  // we load the vector of references
  std::vector<Reference> expected_references;
  ref_dataset.read(expected_references);

  // we use the stored reference and dereference it to gain access in the integers' dataset
  DataSet expected_dataset = expected_references[0].dereference<DataSet>(my_group);

  // as usual, we load the vector with numbers from the extracted dataset
  std::vector<int> read_data;
  expected_dataset.read(read_data);

  // and voila, the payload we excepted
  for (size_t i = 0; i < read_data.size(); ++i) {
    std::cout << read_data[i] << " ";
  }
}

int main(void) {

  try {
    write_dataset();
    read_dataset();

  } catch (Exception &err) {
    // catch and print any HDF5 error
    std::cerr << err.what() << std::endl;
  }

  return 0;// successfully terminated
}
```

Hope that helps.

Keep up the good work!

**Comment by tdegeus** [comment]
Hi @plusangel . Thanks! Maybe you could open a PR?

**Comment by kerim371** [comment]
Hi,

When I launch this example I get error when trying to write vector of references to a dataset (line `ref_set.write(ref_container);`):
![image](https://user-images.githubusercontent.com/43808863/104713759-5e2fb480-5735-11eb-8e80-c34f13df795b.png)

I'm able to write it by `ref_set.write_raw(ref_container.data());` but when I open the dataset in HdfViewer I can see that reference dataset contains single value = 0.

Also I cannot read the data so I have to comment the line `read_dataset();`

I guess the problem is not in the example but rather in something else.

I'm on WIndows 10, hdf 1.12.0 shared library

**Comment by kerim371** [comment]
I've just tested three compilers on Windows 10: MSVC 2019, MinGW, CLang

- MinGW and Clang works fine with this example
- MSVC 2019 gives the error above

I think there is some bug when IO references on MSVC

**Comment by plusangel** [comment]
It might be, I use Ubuntu 18.04.5 LTS x86_64 and GCC: 7.5.0

**Comment by kerim371** [comment]
There is a typo in `H5Converter_misc.hpp` on the line 412:
```cpp
_vec_align.reserve(compute_total_size(_dims));
```
I changed it to `resize()`:
```cpp
_vec_align.resize(compute_total_size(_dims));
```
and now it works on Windows.

Should be fixed in #400

**Comment by alkino** [comment]
Fixed by#396, thanks for all!
